// @small-tech/auto-encrypt/lib/IPAddresses.js
/**
  Gathers external IPv4 address information using https://ip.small-web.org/

  And Ipv6 information using built-in `networkInterfaces` function from node:os.
*/

import os from 'node:os'
import dns from 'node:dns'
import { execSync } from 'node:child_process'
import Throws from './util/Throws.js'

const throws = new Throws()

/**
  @typedef {{ ip: string }} IpAddressApiResponse
*/

export default class IPAddresses {
  /** @type {IPAddresses} */
  static instance

  static extractIPv6Address = "sed -n 's/.*inet6 \\([a-f0-9:]*\\).*/\\1/p'"
  static platformSpecificStableIPv6DiscoveryCommands = {
    'linux': `ip -6 address show scope global | grep -v 'temporary' | ${IPAddresses.extractIPv6Address}`,
    'darwin': `ifconfig | grep 'inet6' | grep 'secured' | grep -v 'fe80:' | ${IPAddresses.extractIPv6Address}`
  }

  /** @type {string|undefined} */
  ipv4Address

  /** @type {Array<string>} */
  ipv6Addresses = []

  //
  // Factory method access (async).
  //
  static isBeingInstantiatedViaAsyncFactoryMethod = false

  static async getInstanceAsync () {
    if (this.instance !== undefined) {
      return this.instance
    }

    // Create singleton instance.
    this.isBeingInstantiatedViaAsyncFactoryMethod = true

    this.instance = new IPAddresses()

    // Initialise it asynchronously.
    await this.instance.discoverIpAddresses()

    return this.instance
  }

  get hasIPv4Address () {
    return this.ipv4Address !== undefined
  }

  get hasIPv6Addresses () {
    return this.ipv6Addresses.length > 0
  }

  //
  // Private.
  // 

  constructor () {
    // Ensure singelton access.
    if (IPAddresses.isBeingInstantiatedViaAsyncFactoryMethod === false) {
      throws.error(Symbol.for('MustBeInstantiatedViaAsyncFactoryMethodError'), 'IPAddresses')
    }

    // This reverts IP address sort order to pre-Node 17 behaviour.
    // See https://github.com/nodejs/node/issues/40537
    dns.setDefaultResultOrder('ipv4first')

    IPAddresses.isBeingInstantiatedViaAsyncFactoryMethod = false
  }

  async discoverIpAddresses() {
    await this.discoverIPv4Address()
    this.discoverIPv6Addresses()
  }

  async discoverIPv4Address () {
    try {
      const response = await fetch('https://ip.small-web.org/json/')
      if (response.status === 200) {
        const ip = /** @type { IpAddressApiResponse } */ (await response.json()).ip
        if (ip !== '::1') {
          this.ipv4Address = ip
        }
      }
    } catch (error) {
      console.warn('Could not discover external IPv4 address.', error)
    }
  }

  /**
    Runs platform-specific command for retriving stable IPv6 addresses as we don’t
    have a built-in way of differentiating between stable and temporary addresses
    using Node.js alone (`os.networkInterfaces()` does not have that information.)
  */
  discoverIPv6Addresses () {
    const platform = os.platform()
    const platformSpecificStableIPv6DiscoveryCommand = IPAddresses.platformSpecificStableIPv6DiscoveryCommands[platform]

    try {
      const stdout = execSync(platformSpecificStableIPv6DiscoveryCommand, { encoding: 'utf-8' })
      this.ipv6Addresses = stdout
        .split('\n')
        .map(address => address.trim())
        .filter(address => address.length > 0)
    } catch (error) {
      console.warn(`[Auto Encrypt] Failed to discover IPv6 addresses using platform-specific (${platform}) command:`, platformSpecificStableIPv6DiscoveryCommand, error)
    }
  }
}
