// @small-tech/https/index.js
/**
  @small-tech/https

  Provides isomorphic access to Auto Encryt and Auto Encrypt Localhost, creating the correct server (localhost using local certificate authority or globally-reachable using Let’s Encrypt TLS certificates).

  Copyright ⓒ 2019-present Aral Balkan, Small Technology Foundation
*/
import os from 'node:os'
import process from 'node:process'
import path from 'node:path'
import EventEmitter from 'node:events'

import AutoEncrypt from '@small-tech/auto-encrypt'
export { IPAddresses } from '@small-tech/auto-encrypt'
import AutoEncryptLocalhost from '@small-tech/auto-encrypt-localhost'

class Events extends EventEmitter {
  CREATING_SERVER = Symbol()
  SERVER_CREATED = Symbol()
  SERVER_CLOSED = Symbol()
}

export const events = new Events()

const AUTO_ENCRYPT_STAGING_SERVER_TYPE = 1

const DATA_HOME = path.join(
  process.env.XDG_DATA_HOME || process.env.HOME || os.homedir(),
  '.local',
  'share'
)

export const SMALL_TECH_HOME_PATH = path.join(DATA_HOME, 'small-tech.org')
export const DEFAULT_SETTINGS_PATH = path.join(SMALL_TECH_HOME_PATH, 'https')

export default class https {
  /** @type { import('@small-tech/auto-encrypt-localhost').AutoEncryptedLocalhostServer| import('@small-tech/auto-encrypt').AutoEncryptedServer} */
  static server = null

  /**
    Creates new AutoEncryped or AutoEncrypedLocalhost https server (asynchronous).

    @param {import('./index.d.ts').SmallTechHttpsOptions | import('node:http').RequestListener} [options]
    @param {import('node:http').RequestListener} [listener]
  */
  static async createServer (options = {}, listener) {
      // There can be only server per node process. Enforce that here.
      if (this.server !== null) {
        console.warn('Create server called when server already exists for this process. This is likely a coding error. Returning existing server instance.')
        return this.server
      }
      
      // The first parameter is optional. If omitted, listener should be passed as the first argument.
      if (typeof options === 'function') {
        listener = /** @type {import('node:http').RequestListener} */ (options)
        options = {}
      }

      const localDomains = ['localhost', 'place1.localhost', 'place2.localhost', 'place3.localhost', 'place4.localhost']
      const domainsListDoestExistOrIncludesALocalDomain = options.domains == undefined || options.domains.some(domain => localDomains.includes(domain))
      const automaticIPAddressDetectionRequested = options.ipv4 || options.ipv6

      const isLocal = domainsListDoestExistOrIncludesALocalDomain && !automaticIPAddressDetectionRequested

      const serverScope =  isLocal ? 'local' : 'global'

      const settingsPath = options.settingsPath ? path.join(path.resolve(options.settingsPath), serverScope) : path.join(DEFAULT_SETTINGS_PATH, serverScope)
      options.settingsPath = settingsPath

      if (options.staging) { options.serverType = AUTO_ENCRYPT_STAGING_SERVER_TYPE }
      delete options.staging

      const messageSuffix = {
        local: 'at localhost with locally-trusted certificates',
        global: 'with globally-trusted Let’s Encrypt certificates'
      }
      events.emit(events.CREATING_SERVER, `Creating server ${messageSuffix[serverScope]}.`)

      const server = await (serverScope === 'local'
        ? AutoEncryptLocalhost.https.createServer(options, listener)
        : AutoEncrypt.https.createServer(options, listener))

      function emitCloseEvent() {
        events.emit(events.SERVER_CLOSED, 'Server closed.')
      }
      server.on('close', emitCloseEvent)

      events.emit(events.SERVER_CREATED, 'Created HTTPS server.')

      return server
  }
}
